### ISSUE TITLE
Option to add own LLM (OpenAI , Anthropic, etc.) API key

### ISSUE THREAD

**Issue by MGNTechVault** [warpdotdev/Warp#2788]
### Describe the solution you'd like?

```text
To have the option to the new Warp AI to add our own API Key and maybe options to tweak the settings like the temperature?
```

### How important is this feature to you?

1 (Not too important)

### Warp Internal (ignore) - linear-label:770f6576-d6c0-4e4f-a259-fc64b5156087

None

**Comment by dannyneira** [comment]
Thanks for this feature request! AFAIK, we plan to keep the Warp AI in preview with 100 calls per day and setup a paid version in the future, so there may not be an option to use your own key but I'm sure it is something that may be possible for enterprise teams/users.

Please :+1:  and subscribe to this request as it helps us gauge interest. We will post any updates here.

**Comment by mf-ms** [comment]
@dannyneira While I can somewhat understand it from a business perspective, I think adoption could be higher if you allow individual users to use their own API keys.
I'd be interested in this feature and even willing to pay to use it (within reason).
Thanks!

**Comment by numist** [comment]
I already pay OpenAI for API access, seems best for all involved if Warp doesn't have to pay any vendors for my AI usage (and my limit goes from 100 to whatever my credit card can handle).

**Comment by HerbWarren** [comment]
Same boat -- I work for a company that's given developers like me API keys to use, and for personal reasons I have my own anyway. If there's an option to bring your own key, that'd be super-hot.

**Comment by hekk-tatsukichi-yamashita** [comment]
Considering its use in a business setting, it would be very helpful to have the flexibility to use customized Azure APIs.

**Comment by Ciel-MC** [comment]
Seconding this, I really, really want to avoid subscriptions if possible, Id be happy to even have an alternative credit system(of course the best case is if it doesn't have to reach out to Warp's server at all and makes the OpenAI call entirely locally, but has a lot of implications), one major reason would be that I would rarely need to use it. I understand that some people using it less is one of the reasons that subscriptions are profitable(and why every company is flocking towards it), but, at least personally speaking, unless there's something else that Id pay for a subscription anyway, I would never pay for subscription just for integrating AI into the terminal, especially when I already have OpenAI credits. This might be a boomer take, but I just want to buy my stuff and have it, damnit.

Sorry if this is a ramble, but I do like the idea of Warp a lot, but if Im going to use a proprietary terminal, it just has to be so, so, much better to be justified, and having to have a subscription for this is a dealbreaker.

**Comment by ThinkFar** [comment]
Please do this.

**Comment by Flecart** [comment]
Is this coming?

**Comment by eighteyes** [comment]
Ping, especially if you start to offer other models, the ability to use our own keys would be nice, for you especially since you're footing the AI bill.

**Comment by YourTechBud** [comment]
Maybe if we have support to override some additional parameters like the base url? That would be dope.
We run our own open source models on a on-prem workstation. The server is openai compatible. Just need a way to change the base url option.

**Comment by nyteshade** [comment]
Companies that want to force using their AI intergration with a pay wall are missing out on the opportunity for a low cost subscription where users can bring their own LLM keys to the party. 

For example you can either get 0 from me for this or 4.99 a month paid upfront once a year if I can provide my own keys that I also already pay for

**Comment by Patcybermind** [comment]
yeah i dont think a lot of people are gonna be paying for a warp subscription since its nothing a little google search and something like perplexity ai or You cant do

**Comment by Flecart** [comment]
Is this coming?

**Comment by Patcybermind** [comment]
Is it planned someday?

**Comment by Flecart** [comment]
> Is it planned someday?

Hope they listen to us...

**Comment by Lr-2002** [comment]
Yep, this will limited your company's income which is really terriable, But could we just share a mode with cheaper subscription which could give an interface to accept the personal api token ? I would pay for it for the most modern terminal i've used .
It's really awesome to use warp terminial and it helps me a lot .Hoping for your reply

**Comment by DanEdens** [comment]
This new model totally ruins the terminal for me. I'd have happily paid 5, and begrudged 10, but 22 is absurd. I can just use the chatgpt-cli for virtually the same result.. 

There really needs to be a personal tier. Even calling it a team when I only intend for myself, feels like getting a veggie option and still having to pay for the meat

**Comment by Patcybermind** [comment]
yall do know that github copilot cli exists and its free for a lot of people

**Comment by Patcybermind** [comment]
we arent gonna pay 22 a month for some extra gimmicks

**Comment by ShivamB25** [comment]
even a one time fees or yearly would have been fine with Bringing our own keys. Warp isn't so so truly open source.

**Comment by Patcybermind** [comment]
Its as open source as google chrome

**Comment by holdjun** [comment]
Please do this.

**Comment by Mjb141** [comment]
Is this still not a thing? You're going to lose a _ton_ of corporate potential clients if you don't allow us to bring our own API keys.

**Comment by eighteyes** [comment]
One Agent session eats up 1/8th of your AI credits for a month. No thanks, I'm going to find another terminal that doesn't ask me to subscribe to overpriced AI Arbitrage ( such a innovative biz model ) every time I use it. 
![image](https://github.com/warpdotdev/Warp/assets/405274/328b69fb-de64-4520-9c0a-28308c163520)

btw: I've been on iTerm's Patreon for years now, I support good dev tooling. They didn't do the best AI implementation, but at least it's bring your own key. 

You know what I do with AI? I don't just let it run commands on the prompt, especially if it's a complex workflow. I save the steps in a text file so I know what I did and I can reproduce it later if need be. I'm already copying and pasting. 

`#` is nice because it is non-intrusive, and 100 requests is fair ( 40 is annoying ). Agent is pushing it in our face and cluttering up the terminal with advertisements. Terminal is a sacred place between a person and machine, you just injected your business into it, and I'm removing your hotkey for that.

**Comment by zsxkib** [comment]
Is this a thing yet?

**Comment by NebulusIO** [comment]
If they don't then just wait for the hype to die down.

**Comment by dolbyzerr** [comment]
Well, it was nice while it lasted, I was contemplating to get a subscription. But 15$ (18$ monthly) per month? I would probably be ok with pay as you go (like 0.01$ per request). 

Don't get me wrong, Warp is good product, I do understand that it's also a business, I am happy to support a good product but it has to be competitive and provide enough value for me (and it seems I am not alone in this).

Aaaand back to iTerm I go

**Comment by nixolas1** [comment]
Was hyped when I tested out agent-mode today, with some advanced JQ JSON manipulation, but ran out of tokens before I could complete the task.
But paying $15-20 for more than 2 AI tools a month is just not possible. Would have to switch out ChatGPT or Copilot with Warp, which is much more niche.
Would love for you to find a product model that works for you, Warp Team, while still being attractive!

**Comment by GGxsilverx** [comment]
please consider doing this , we will eventually find a workaround for this if you don't , i know that you might thinking on have some monopoly around this feature or use it to train something , but believe me nobody was successful with such plans and you will be no exception and this might be a feature that some other product compete with you or you just simply loose your hype against some big names like iterm.
but if you open this route you might get surprised by creativity and techniques people add to your platform.

**Comment by zed-wong** [comment]
I don't think this will be added, maybe the community can fork it and add it in a branch

**Comment by xhlr8** [comment]
Given how old this issue is and that many relevant other posts have been made for this simple request, I doubt this development will ever bring their heads out of their *sses. An otherwise attractive product ruined by the priority need for a redundant business model.

**Comment by rexeo-asia** [comment]
check in for October 2024. Wake me up when 2025 arrives

**Comment by randybias** [comment]
+1000

**Comment by fef5002** [comment]
up-voting this request, because I couldn't figure out how to upvote.  I also would love to have speech to text capabilities within Warp but I'm not sure how one goes about adding that...

**Comment by iiprisma** [comment]
+1.
This terminal is great, but I would switch in a heartbeat to a terminal with half the features but which has the ability to bring your own api key. 

Well, we'll see what happens

**Comment by fantasyxby** [comment]
Take a look at cursor, a good example from the IDE.

**Comment by fantasyxby** [comment]
Take a look at cursor, a good example from the IDE.If you guys want warp to be as hot as cursor, do it!

**Comment by Qervas** [comment]
Yeah, same request here. I'm a fan of Zed Editor, they support popular API keys and local LLM model interfaces like Ollama.
Just more options for different personal flavor preferences. If that day comes, this terminal would be dope w/o question.

**Comment by vfilter** [comment]
So much hate, abuse and entitlement here. Guys, you're getting a great terminal for free. How much use is it to shit on a hard working team of devs here? You want the feature, you pay for it. Otherwise, just use one of many alternatives, it's really that simple.

**Comment by vfilter** [comment]
@rosmur I see where you're coming from, but that was not the point I was trying to make. The comments in this thread stink of entitlement and pettiness. It's fine to give feedback, it's fine to say that you leaving for iterm is a decision you made based on a decision the Warp team made. But saying things like the devs should "get their heads out of their asses" is simply uncalled for.

**Comment by bugs181** [comment]
> So much hate, abuse and entitlement here. Guys, you're getting a great terminal for free. How much use is it to shit on a hard working team of devs? You want the feature, you pay for it. Otherwise, just use one of many alternatives, it's really that simple.

My 2 cents here. I'd gladly keep paying for Warp, but still want the option of using OpenAI. That's because I do a lot of self-hosting and have my own open-ai compatible API that are also plugged into a lot of other systems (logging, analytics, and cross-device/cross-app capabilities. MAYBE I am in the minority here and don't see this as entitlement. I have thought about cancelling Warp subscription, primarily for this reason. And it's one less sub to worry about.

What I'd REALLY love to see, is a one-time cost piece of software (say a year's worth of updates) - and the ability to bring your own openAI. I know the trajectory of all things in life is the subscription model - but I'm just not on board with it. If I were to subscribe to EVERY single piece of tech I use, I'd be paying thousands - it's just not gonna happen. It's not plausible. I doubt I'm alone in feeling the economic pressure and just want to go back to a simpler time. You get what you pay for.

**Comment by rosmur** [comment]
> @rosmur I see where you're coming from, but that was not the point I was trying to make. The comments in this thread stink of entitlement and pettiness. It's fine to give feedback, it's fine to say that you leaving for iterm is a decision you made based on a decision the Warp team made. But saying things like the devs should "get their heads out of their asses" is simply uncalled for.

TBH, im fine with it. If someone wants to bring SaaS subscription bullshit to something as fundamental as the terminal. The software subscription business model is an absolute scourge and needs to be vehemently opposed. Furthermore, if you make a Github "repo" with no actual code, just to "engage users", this is exactly the kind of treatment I expect them to get. In fact Im happy the community is reacting this way to keep companies such as warp in check - lest they get even more grubby.

**Comment by poisson-fish** [comment]
yeah i'll just go ahead and use iterm2, plug in an openai API key and avoid the middleman. shame, i wanted to like warp

**Comment by BryanJBryce** [comment]
@poisson-fish Going the same route 🖖🏼

**Comment by AlexandreSenpai** [comment]
I just wanted this feature cuz I dont want to pay again for something that I already have lol but I do understand their point of not doing it. I'm going to keep using warp because its very good and I personally don't use AI features so much.

**Comment by jonathansampson** [comment]
I found Warp today (via https://youtu.be/qkduRen6QFk) and immediately saw the value. I was, however, quite disappointed to learn that I cannot connect it to _my own local models_. Connecting a third-party with my system/input/output in this manner is not ideal for security and privacy. Permitting the use of local models would dramatically increase Warp's attraction for the cohort of users who wish to keep things locked down and private.

**Comment by MistressPlague** [comment]
Please allow local models, such as allowing us to use the openAI like LM Studio API.

**Comment by SebastianAtWork** [comment]
Also from me, I need this feature. Since i live in the EU, I cannot use this AI for Data Protection Reasons. Custom Api Key would solve this.

**Comment by GoDjMike** [comment]
**UPDATE**: I emailed Warp asking if I could bring my own API key. A support engineer from the Warp team shared this in their response:

> Hey Mike,
> We are tracking it in this GitHub request to be able to use your AI API key here: https://github.com/warpdotdev/Warp/issues/2788
> 
> _**We're actively investigating this feature and its implementation, billing, etc. Please 👍 the issue to show interest and subscribe as we'll be putting all updates in the GitHub thread above.**_
> 
> Let us know of any issues, feedback, or questions.
> Danny from Warp, Support Engineer

**Comment by LaurentiuSimionescu** [comment]
I’m in the same boat here. I use Warp and have already paid for Copilot and ChatGPT, plus I run Ollama locally. Adding yet another subscription is just unreasonable—it’s the same reason I stopped using Replit.

This feels like the subscription wars all over again—Netflix, Disney+, Apple TV, Amazon Prime. At least with them, I get why—they have licensing costs and content deals to manage.

But you have the chance to take a different path. Warp can be the best terminal by listening to its users. There’s a high demand for this feature right now—don’t let rigid monetization be the roadblock. I’m more than willing to pay a subscription as long as I can use my own LLM. Give users that flexibility, and you’ll have a stronger, more loyal community.

**Comment by mochetts** [comment]
Charge me to let me use local models / own api keys removing the limit in requests. I'd gladly pay.

**Comment by iAbadia** [comment]
I would gladly pay for your development, just don't make me pay for the use of comercial LLMs that I'm already paying for.

**Comment by oscarmorrison** [comment]
+1

**Comment by mikenerone** [comment]
Many users have shared valid personal reasons for needing to use their own API keys. From a business perspective, the enterprise use case is perhaps even more critical. My company, like many others, has established data protection agreements with several AI providers. While I recognize that Warp's utilization of OpenAI's APIs currently prevents proprietary information from being exposed through training, without a formal legal agreement, this amounts to a "pinky promise" that could be altered at any time. For this reason, it's simply not viable for our organization.

**Comment by mikenerone** [comment]
@mochetts Thanks I didn't realize the enterprise plan included that. I guess what I'm facing is more of a chicken-and-egg problem then. Without the ability to individually use API keys, we can't get to a critical mass of individual devs using Warp for it to make sense for the company to consider upgrading to that.

**Comment by mochetts** [comment]
@Mjb141 @mikenerone Disclaimer: I don't have any relation with warp other than being a user.

I believe you could build a case within your company by trialing the Turbo plan (30 usd/user/mo), or even individual Pro plans, and exposing the benefits of using within your team of 30. 
I'm certain the warp team will give you a good enterprise deal for 30 seats (maybe you can negotiate something under 15 bucks / month if you bring your own llm).

IMHO, companies should not be the ones having issues as there are plenty of options to build a case for an enterprise plan.
Individual users (freelancers or contractors) are the ones struggling with the price.

**Comment by anubhav94N** [comment]
First of all, this is a kickass product, and I absolutely love it. It is my first day using Warp, and I rammed through half my monthly quota in the Free Tier in a couple of hours! I feel like my terminal skills are on steroids when using Warp.

I understand that Warp is a business and needs to make money. At the same time, the product raison d'etre (to me) seems to be increase developer productivity, and make every user feel like a terminal wizard. To that end, I fully agree with the sentiment expressed in so many of the comments here. I would happily pay 10$/month to Warp for just using their product, but being able to connect my own LLM (either self-hosted; or by providing a vendor's API key). Reading this thread, I think there are many users who might consider such a pricing model; so I suspect there's a business case to be made that such a pricing model will in fact increase Warp's revenue. I hope the Warp team will consider this feature request.

Either way, thanks for building such a cool product.

**Comment by dataPhysicist** [comment]
I'd like to change the base URL & API key so we can use [virtual API keys](https://subkeys.link/j34WNaU)  (security and overage protection).

**Comment by jSayal** [comment]
I will start using Warp as my main terminal, if I use my own API keys. I think Warp is missing a lot of users by not providing this feature. Sooner or later some other open source terminal emulation project will start providing this feature and a lot of users will move to it.

**Comment by randybias** [comment]
Wave Terminal.

**Comment by jbadenho** [comment]
Warp is already redefining what a terminal can be, and support for custom LLM API keys would cement its position as one of the leading AI-native terminals.

This request goes beyond cost savings or user flexibility. It’s about unlocking a broader developer experience where Warp becomes a canvas for deeply personalized, context-aware automation. Letting users plug in their own models or API keys lays the foundation for:

- Project-aware AI agents that understand local context  
- Seamless integration with private/self-hosted LLMs (Ollama, LM Studio, Azure, etc.)  
- Advanced workflows tailored to individual developer environments

This level of extensibility would attract not only power users, but also teams and open-source communities looking to create custom workflows and intelligent tools that run *with* Warp as the interface. It positions Warp not just as a standalone terminal, but as a platform others can extend—building their own automations, agents, and developer experiences *on top of* Warp’s foundation.

You’ve built an incredible product. Supporting user-defined LLM endpoints could help Warp become a **highly extensible, AI-native dev platform that feels like home for all your daily workflows**.

At a higher level, this feature enables a collaborative model where it’s no longer just the developer working alone. Instead, it becomes a team of three:
- The **developer** (you),
- **Warp**, an intelligent and project-aware AI interface that understands context and acts with precision,
- And an **LLM** like OpenAI, Anthropic, or a self-hosted model providing creativity, analysis, and insight.

With this trio working together as technical partners, developers of all skill levels could build faster, solve problems more confidently, and stay in flow. That’s not just convenient—it’s transformative.

Hope this lands on the roadmap in a meaningful way. Thanks again for pushing terminals forward.

**Comment by HerbWarren** [comment]
To piggy-back on @jbadenho said (and thanks much for that!), as AI technologies like MCP grow, it's far more likely than not that developers are going to have their own MCP servers available to provide context to their projects, and being able to tap into those through the terminal experience would be amazing. 

Let Warp _use_ whatever AI the user wants to bring to bear -- this world is changing so fast that attempting to lock devs into tooling isn't going to be viable IMHO. Open beats closed, and having an amazing terminal experience that people can extend to use their AI tools of choice is going to make Warp more valuable to devs of the future (including today) than a terminal that merely "provides" AI tools.

**Comment by Death916** [comment]
+1 for this. Would be a killer feature. Just add a bring your own tier price or something.

**Comment by abnerfreitas** [comment]
+1
Why are we still discussing this?

**Comment by ibussieres** [comment]
+1.

0% chance I pay another AI subscription for this. But I would pay a decent subscription for a terminal that let's me use my own AI provider keys.

**Comment by Dxuian** [comment]
why isnt this implemented yet ?

**Comment by vaisov-gemba** [comment]
> Wave Terminal.

That's a nice alternative! Though it doesn't have agent mode :(

**Comment by kealuya** [comment]
Alternative tools can be used, such as mods, which allow you to use custom LLM keys in Warp.

**Comment by vaisov-gemba** [comment]
> Alternative tools can be used, such as mods, which allow you to use custom LLM keys in Warp.

Sounds amazing, could you direct me to such mods? Can't find them

**Comment by kealuya** [comment]
> > Alternative tools can be used, such as mods, which allow you to use custom LLM keys in Warp.
> 
> Sounds amazing, could you direct me to such mods? Can't find them

https://github.com/charmbracelet/mods        
Hope you like it

**Comment by vaisov-gemba** [comment]
> 
> https://github.com/charmbracelet/mods Hope you like it

Not sure I'm following, but how this mod can enable custom LLM keys in warp terminal?

**Comment by mochetts** [comment]
@vaisov-gemba you wouldn't be using warp to do LLM completions. You would be using this mod tool. It's not bad tbh.

**Comment by MacJediWizard** [comment]
how does this work as i have my own llm i could use with warp to save on tokens. I would really love to know

thanns



William Grzybowski

MacJediWizard Consulting, Inc.
971 US-202 North
Suite N
Branchberg, New Jersey 08876
***@***.***
(973) 310-2345


￼￼

**Comment by vaisov-gemba** [comment]
> [@vaisov-gemba](https://github.com/vaisov-gemba) you wouldn't be using warp to do LLM completions. You would be using this mod tool. It's not bad tbh.

but does it have agent mode? looks like not.

**Comment by arunsr1ni** [comment]
This is stopping me from using warp. Would love a way to use my own keys and LLMs. Easier to get enterprise access as well.

**Comment by jakubsuchybio** [comment]
I understand the warp team not priotizing this as it could undercut their funding.
With that said I agree with an option to pay under $5 fee for using other LLMs through API keys, so that they are not cut off completely.

**Comment by l4kr** [comment]
If you give me a 50% discount on your AI services and also reimburse 50% of my OpenAI subscription I'll go for it.

**Comment by pfurini** [comment]
Please prioritise this, at least the ability to use local models, for example via ollama.
Not everyone is willing to send everything to 3rd party servers, me and my company included...

Sorry but we're sticking with iterm until you implement BYOK in warp. Nice product otherwise!

**Comment by johnib** [comment]
I don't anticipate Warp team ever fulfilling this request, so if you're up to losing the beauties of Warp, then do the following:

1. Install VSCode and CLINE extension.
2. In CLINE's Act mode -- just tell it what you want to do **exactly as you would have in Warp**, and observe the magic happens.

In CLINE you can just plug it to whatever API key you already have.

@dannyneira it may be interesting for you folks to look at CLINE's model -- they offer their own LLM provider "cline", but they let you bring your LLM easily.

I am a 3+ years Warp user. And this missing functionality has made my Warp usage decline almost to zero.

**Comment by Bnjoroge1** [comment]
+1 on this. Anyone know what a great alternative is that has Agent mode or something similar?

**Comment by vaisov-gemba** [comment]
> +1 on this. Anyone know what a great alternative is that has Agent mode or something similar?

I actually ended up using Cursor AI (our company bought us the team plan) with Sonnet 4. It can log in to servers and do the same stuff Warp can. And it's even better that it has access to your code so you can do both. It's just the interface isn't so terminal adapted, but it does the job. I'm pretty sure Cursor will implement at some point some kind of terminal management and then Warp will become useless.

Also I can see a business opportunity here - build similar tool as Warp, sell it with 10-15$ subscription but with ability to use your own LLM keys, that's it, market is yours.

**Comment by jbadenho** [comment]
This gives me a feeling of moving beyond the “one tool to rule them all” mindset and starting to think about AI orchestration—letting different AIs play to their strengths in the dev workflow. Using Warp in the terminal, Cursor in the editor, and ChatGPT for code review/specs/brainstorming. Each fills a different niche and swapping out other AI's for flavour:

Warp gives AI versatile use in the shell,

Cursor (and similar IDE agents) actually build, refactor, and understand whole projects,

Foundation models like ChatGPT help step back, review, and get clarity.

The magic comes from connecting these strengths—using each tool for what it does best, and handing off context between them as one moves from idea to code to deployment.

That’s why I also want Warp to be open to different LLMs, let's “bring our own key,” or even better—collaborate more closely with other best-in-class AIs. Of course, this vision also means being very mindful of privacy and compliance—ensuring that workflows can cater to users and teams who need to keep sensitive data in-house, or require strict control over where information flows and which LLMs are allowed to access it.

If a workflow where Warp, Cursor, and ChatGPT could actually share context and build on each other’s outputs—securely and in line with company or personal data policies—making the full coding lifecycle (from first spec to running code) seamless, “remembered” at every stage, and compliant with even the strictest requirements.

What if AI tools embraced an orchestral, collaborative approach—no longer siloed as gatekeepers to their specific domains, but working as a unified hub that lets developers mix and match their favorite tools, all while ensuring complete data control and compliance?

Curious what others think:

How are you all mixing and matching AI in your current workflows?

What would a integrated, project-aware coding environment look like now—especially for those working with sensitive or regulated data?

Let’s keep striving for dev tools—and teams—that truly work together, empowering not just seamless human-to-human collaboration, but also enabling humans and AIs to team up effectively. And let’s make sure these tools remain affordable and safe—so even small teams, startups, and independent developers can thrive, create jobs, and build sustainable livelihoods in this new AI-powered landscape.

**Comment by gnorbsl** [comment]
I started looking for alternatives depsite loving the product, I understand they want to make money but all I want is an agentic AI in the terminal I do not need all the fluff around it that inflates their costs and I'm not willing to pay 15 for another service if I'm already subscribed to LLM providers. 15 is just too expensive for what I need or use.

**Comment by Amrree** [comment]
This needs to work with ollama. Either built-in access or MCP access, but actively withholding this is wrong. Someone with a paid subscription should be able to hook up local, private llm.

**Comment by Mjb141** [comment]
> https://blog.google/technology/developers/introducing-gemini-cli-open-source-ai-agent/

Thanks for this, installed!

**Comment by vaisov-gemba** [comment]
> https://blog.google/technology/developers/introducing-gemini-cli-open-source-ai-agent/

Thanks! Amazing! Just installed and tried it already - works perfectly for my needs, gonna cancel warp subscription :)

**Comment by pasquale95** [comment]
Very sad to see that Warp doesn't allow to use local LLMs or your-own-API-Key. You have everything in place to settle as the standard-de-facto terminal on all the 3 platforms, it's a pity that you're missing out such opportunity by forcing users to use your paid plans when, either for company restrictions or simply for cost savings users would like to use their own LLMs.
Alex Ziskind said in his video that you've in plan to allow users using their local LLMs. Hope this is true, it would help catching with waveterm.

**Comment by jzumwalt** [comment]
As a third-party software consultancy, it's hard enough to convince customers that OpenAI, Anthropic and Google aren't stealing their source code. Not offering a version of Warp that doesn't route requests through your servers means it's no-starter for our company (and I'm assuming many like ours). I'd still be willing to pay per seat, as an outsider, this seems like free money for Warp?

**Comment by bdmorin** [comment]
This is critical to adoption. I don't know why anyone would push their console data through your servers.

**Comment by vaisov-gemba** [comment]
these are the tools you can use instead of Warp

- Claude Code
- Gemini CLI
- Qwen Code

**Comment by Fgruntjes** [comment]
This would also be needed to be using https://groq.com/ for way faster and cheaper inference.

**Comment by UsmanTechJive** [comment]
I am super excited about this feature. Please bring it on!

**Comment by christian-wa** [comment]
I'd be happy to pay for the option to set an own API key and URL as well.
Our company requires the use of a gateway for LLM access so not having the option to set an own URL and key is basically a direct no for a tool to get approved.

**Comment by muyideensakazakariya-eng** [comment]
> ### Discord username (optional)
> 
> _No response_
> 
> ### Describe the solution you'd like?
> 
> ```text
> To have the option to the new Warp AI to add our own API Key and maybe options to tweak the settings like the temperature?
> ```
> 
> 
> ### Is your feature request related to a problem? Please describe.
> 
> _No response_
> 
> ### Additional context
> 
> _No response_
> 
> ### How important is this feature to you?
> 
> 1 (Not too important)
> 
> ### Warp Internal (ignore) - linear-label:770f6576-d6c0-4e4f-a259-fc64b5156087
> 
> None

**Comment by 2214372851** [comment]
I can feel that Warp Team does not pay so much attention to users, we have given a reasonable win-win plan and launched a separate LLM custom package, but it has not been solved from 2023~2025, I don't think this is the attitude that a software should have towards users.

**Comment by invntiv** [comment]
how the hell is this feature still missing two years later? I like warp but im considering leaving it in the dust after seeing the lack of movement on this. I pay $200 a month for my anthropic key. If warp's only business model is being a middleman to resell me api credits through their own subscription service, I'm not interested. give me a $4.99/month option to help you sustain your business and I'll gladly do that to be able to put in my anthropic key. what I'm NOT going to do is support this growing trend of every app under the sun now becoming a reseller of api credits that they buy in bulk.

**Comment by mymark21** [comment]
please ship BYOK feature!

**Comment by christian-taillon** [comment]
## Current State

It appears the "bring your own API key" option for Warp AI is currently only available under their [Enterprise plan](https://www.warp.dev/pricing).

## Use Cases & Challenges

### Small Business Perspective
As a small business owner, I find the current pricing prohibitive. I'm a Pro customer and would appreciate a more accessible option. Many organizations permit AI usage only with local models, and solutions like `vllm` or `OpenWebUI` are used to expose coding models. Some users have noted that using their own models with Warp avoids Warp's costs, which is a common approach for those reselling API credits.

### Enterprise Requirements
An enterprise I work with has **limited pre-approved LLM provider options** due to security and GRC requirements. A **BYOK** (Bring Your Own Key) or **BYOLLM** (Bring Your Own LLM) option would enable them to utilize this tool.

## Recommendation

For those considering this, **engaging with the sales team** is a current avenue. I'm willing to pay for the product and use my own models, which would reduce Warp's expenses. However, the current enterprise pricing is not feasible for many small businesses and their developers, who typically use more affordable monthly subscriptions.

### Proposed Solution
Introducing a "bring your own LLM" option, even with limitations like:
- Team size caps
- Revenue limits
- Limited model providers
- Usage quotas

This could significantly boost Warp's adoption across different market segments.

**Comment by diegobayshanntech** [comment]
Heard good things about Warp. So I came to see if I can use our Anthropic API Key. This is a must policy with our company in order to protect our IP (Anthropic Teams plan does not use data for training). But found this sad thread.

Bye again, was nice meeting you.

**Comment by HelloKitty921** [comment]
There must be one butthurt fool up the 3-step ladder not approving this. Would rather see the entire product fail, than give users what they want. Sad

**Comment by ProjectInfinity** [comment]
Add BYOK, even if only for paying customers (similarly to Cursor). There are many models out there and it is impossible for you to support them all, therefore the easy choice is to let us use OpenAI compatible APIs such as openrouter.

**Comment by 2b3pro** [comment]
> Yes, I would absolutely pay $5/month to get to BYOK.

Same here!

**Comment by tpiros** [comment]
I hate to ask this but do we know an alternative product option that actually allows to bring your own key?

**Comment by srhardy** [comment]
I too, really love everything Warp is doing but not being able to byok is unfortunately a deal breaker for me! Im on the FREE plan and I wont be change that until I can use BYOK API. Someone will just forkyou and add it, so come on guys!

**Comment by klausagnoletti** [comment]
I have both a subscription to OpenAI, Claude and Gemini. Currently I don't utilize these enough and bying a license for yet another AI tool seems a bit dumb when I don't utilize what I already have. So please add the option to add 3rd party AI tools. I will primarily use it for Code.

**Comment by JordanM575** [comment]
I know the business model is to make money via their subscriptions. However, think of how many people would be using your product if you enabled BYOK. I would be recommending it left and right, and for the people who don't have BYOK capabilities they would sign up if they like your product. Its free marketing and what is stopping your awesome product from being a staple in the devstack.

**Comment by amannirala13** [comment]
The feature is much needed. I am already paying for ChatGPT, Claude, and Perplexity subscriptions, and paying for using Agents through the same models I am already subscribed to is not pragmatic. Considering it has been over two years since it was originally proposed, the chances of this option materializing are bleak.

I believe we need capable people and organisations that are strongly motivated by the idea of open and collective growth to address this entire paradigm regarding how we interact with AI services and the standards and business models we adopt for developing future solutions. 

Until then, I will continue using Wrap as it’s a good terminal and maybe a good AI tool, which sadly I can’t appreciate because of really conservative business decisions.

**Comment by vargalas** [comment]
I would swirch to warp is I could use my existing mmax plans and just pay extra for your service. But my tokens are already paid

**Comment by MadeInShineA** [comment]
Same

**Comment by coorslight** [comment]
we all have maxed out big ai subscriptions. yes, pivoting sucks. but you're gonna lose us w/o BYOK.

**Comment by K3UL** [comment]
> we all have maxed out big ai subscriptions. yes, pivoting sucks. but you're gonna lose us w/o BYOK.

> tbh this. I pay for the 20x Max plan for Claude which would be nice to have an integration option for. [...]

Absolutely same.

Not only this feature would be a good selling point to penetrate the enterprise environments, but it even makes sense as a business decision. Warp should focus on what sets it apart as a TERMINAL, admittedly also on some good AI integration like it seems to have.

But nowadays every other app has a side hustle with a token subscription. And then I need the Claude sub, ChatGPT sub, Warp sub, ... Feels like the streaming providers.

And then if one of the actual model providers (anthropic, openAI, etc.) decide to make a terminal app and cut access for certain competitors like Warp, it will be good to have put more effort into selling a good terminal (like iterm does) instead of riding the hype.

As it stands, personally I would love to try the agents it but I am not going to add another subscription to use the models I can already use with my Claude and Copilot subscriptions.

Warp remains an awesome terminal without these features though

**Comment by vargalas** [comment]
Yeah! Resubscribing now

**Comment by vargalas** [comment]
Nope, guys. I did resubscribe in the hope of being able to use my claude and openai subscription. But not. I can use my api keys. But the most requested feature was that we should be use the subscription. I'm already payin 400Euro for the two, which is able to run also terminal commands  with ai. I would pay the 20$ for warp, it is good if I can use my subscription, but you only added the possibiity of using api keys. It doesn't woth anything for me.

**Comment by regismesquita** [comment]
Same as the others here... we have great models running locally, I would love to be able to have lower latency, higher privacy and just use my self-hosted models, that would be a really nice feature.

**Comment by bholmesdev** [comment]
Thanks for pushing on this over time — the feedback on BYOK has been very clear. @zachlloyd  posted an update: we plan to make BYOK available on the free plan in the next 1–2 weeks for individuals and companies up to 10 employees, and extend it to arbitrary OpenAI-compatible endpoints. Details here: https://github.com/warpdotdev/warp/discussions/9619#discussioncomment-16831279

**Comment by 825i** [comment]
> Thanks for pushing on this over time — the feedback on BYOK has been very clear. [@zachlloyd](https://github.com/zachlloyd) posted an update: we plan to make BYOK available on the free plan in the next 1–2 weeks for individuals and companies up to 10 employees, and extend it to arbitrary OpenAI-compatible endpoints. Details here: [#9619 (comment)](https://github.com/warpdotdev/warp/discussions/9619#discussioncomment-16831279)

This is an absolutely massive win for everyone. Respect to Warp team.
